Make sure that AutoGroups work correctly also when concurrent processes are executed

giovannipizzi opened the pull request:
If many `verdi run` processes were run at the same time, there were cases in which the automatic group name generated by the AutoGroup was conflicting, resulting in IntegrityErrors from the database.
This issue fixes this problem by automatically generating a new name until this is unique.
Moreover, many bugs and issues found while addressing this issue are solved.

Fixes #997

giovannipizzi committed:
Clean-up of the autogroup logic

In particular:

- deprecated use of 'group-name' in favour of 'group-label'
  both in the internal API of AutoGroups and in the CLI parameters
  of `verdi run`
- Improved the Autogroup implementation and API
- add various tests of the functionality of autogroups
- fixed the CLI parameters to include/exclude groups, that were badly
  broken (and untested). Now it accepts any entrypoint string (before
  it was accepting the first part of an node_type string). Also added
  click validation of the parameters.
- fixed the --group flag to activate/deactivate autogrouping (it was
  wrongly defined and thus always True)
- linter fixes

giovannipizzi commented:
I would suggest not to squash the two commits - I am not 100% sure if the first, alone, would pass all tests, but the changes are splitted in two logical units.

If this is an issue, ok to squash, but then please keep both commit messages as they contain relevant information.

giovannipizzi commented:
I've rebased this, so it's ready to review (hoping the tests didn't break in the meantime...). @ramirezfranciscof do you want to go through this (together, or alone)?

Also, @sphuber, looking back I realised that I improved the performance of storing nodes.
After we merge, could you please re-run the test to see how many simulations you can submit at max rate, so we get a feeling if this change already gave an important performance boost? (it might!)

Also, let me reiterate the comment on *not* squashing, if possible.

giovannipizzi commented:
Sorry, it's taking a bit longer than expected as I found a new unexpected problem in the tests for backwards-migrations.
I have identified now the problem (a backward migration wasn't resetting correctly the migration after being run) and I have just force-pushed a commit that should fix this problem and actually also improve the code style of that specific class.
I have also fixed a few more prospector complaints along the way (for some reason unrelated to my changes, but it was complaining locally...).

greschd commented:
Maybe related to the pre-commit and migration issues encountered in #3738?

giovannipizzi commented:
> Maybe related to the pre-commit and migration issues encountered in #3738?

Most probably - let me fix it here (actually it's now fixed, I'm just fixing another hopefully final pylint complaint, that BTW I see was also addressed in #3738.

I've already force-pushed, if it works I'd suggest to merge this first.

BTW, I hate when you fix all problems and then prospector starts complaining of something new... ;-)

giovannipizzi commented:
Tests pass!
![success](https://www.reactiongifs.com/r/cheering_minions.gif)

@sphuber @ramirezfranciscof this is ready for review, and @ltalirz this should also fix the migration test problem (or at least I fixed a problem in the same place you saw problems in #3738)

sphuber commented:
Before I start reviewing, @giovannipizzi I am afraid you may have made a mistake in your rebasing. You now have two commits, but both with the same commit message

giovannipizzi commented:
> Before I start reviewing, @giovannipizzi I am afraid you may have made a mistake in your rebasing. You now have two commits, but both with the same commit message

Oh no... how was the reflog trick to get back old commit messages? Anybody knows?

giovannipizzi commented:
Never mind, I think I found it. I'll force-push with the correct message.

giovannipizzi committed:
Centralised the creation of the Autogroup

This also remove an overzelous isinstance check, and
moves additional checks in a cached function that is run only when
storing the very first node (that needs to be put in an autogroup),
making storing of nodes faster (even if times oscillates so it's hard
to estimate exactly by how much).

Also, added logic to allow for concurrent creation of multiple groups
(and test). This fixes #997

giovannipizzi commented:
Ok @sphuber I fixed the commit message.
I hope I didn't mix up the content of the two commits - in the worst case, ok to squash but keeping text from both commit messages.

sphuber reviewed (changes requested):
Thanks @giovannipizzi . Have some comments, maybe easiest to discuss in person on skype/zoom?

giovannipizzi commented on aiida/cmdline/commands/cmd_run.py:
```diff
@@ -37,31 +40,73 @@ def update_environment(argv):
-@click.option('-g', '--group', is_flag=True, default=True, show_default=True, help='Enables the autogrouping')
-@click.option('-n', '--group-name', type=click.STRING, required=False, help='Specify the name of the auto group')
-@click.option('-e', '--exclude', cls=MultipleValueOption, default=[], help='Exclude these classes from auto grouping')
+@click.option('--group/--no-group', default=True, show_default=True, help='Enables the autogrouping')
```
ok, will put default=`False` (will put a `is_flag`, with name `--autogroup`)

giovannipizzi commented on aiida/cmdline/commands/cmd_run.py:
```diff
@@ -37,31 +40,73 @@ def update_environment(argv):
+    required=False,
+    help='Specify the name of the auto group [DEPRECATED, USE --group-label-prefix instead]'
+)
+@click.option(
```
ok. I'll check in aiida.cmdline for the utility of th calculation/process query builder, for similar functionality

giovannipizzi commented on aiida/orm/autogroup.py:
```diff
@@ -32,148 +34,240 @@ class Autogroup:
+            if (the_class not in exclude_exact_classes and not issubclass(the_class, exclude_with_subclasses_classes)):
+                # If we are here, it's not excluded
                 return True
+            # If we're here, it's both in the include and in the exclude - exclude it
```
I'll make the two flags mutually exclusive

giovannipizzi commented on tests/backends/aiida_sqlalchemy/test_migrations.py:
```diff
@@ -116,8 +120,7 @@ def _reset_database_and_schema(self):
         self.reset_database()
-        with Capturing():
-            self.migrate_db_up('head')
+        self.migrate_db_up('head')
```
I'll put it back

giovannipizzi replied:
Actually - this is only for sqlalchemy, and this doesn't seem to produce any output for me even with the `pytest -s` option - so I would leave as is (i.e. removing `with Capturing`). Ok for you @sphuber ?

giovannipizzi commented on tests/backends/aiida_sqlalchemy/test_migrations.py:
```diff
@@ -57,22 +56,27 @@ def setUp(self):
+            self._perform_actual_migration()
         except Exception:
             # Bring back the DB to the correct state if this setup part fails
             self._reset_database_and_schema()
```
I'll put just for safety

giovannipizzi commented on aiida/orm/autogroup.py:
```diff
@@ -32,148 +34,240 @@ class Autogroup:
+        self._group_label = None
+
+    def get_or_create_group(self):
+        """Return the current Autogroup, or create one if None has been set yet."""
```
ok

giovannipizzi commented on aiida/orm/autogroup.py:
```diff
@@ -32,148 +34,240 @@ class Autogroup:
+        label_prefix = self.get_group_label_prefix()
+        # Try to do a preliminary QB query to avoid to do too many try/except
+        # if many of the prefix_NUMBER groups already exist
+        queryb = QueryBuilder().append(
```
Decision: keep the logic, add comment, add direct tests by calling get_or_create by hand

giovannipizzi replied:
See also my comment above. This function would return 0 both if `labels == []` and if `labels=['prefix']` (actually, I think it would raise when getting the counter because `interegers` would be an empty list).
This can be improved, but I don't think we would simplify too much, we would just shift some logic from SQL to python. For now we keep like this, and we can improve later (e.g. if there are performance considerations).

giovannipizzi commented on aiida/orm/nodes/node.py:
```diff
@@ -1024,18 +1025,10 @@ def store(self, with_transaction=True, use_cache=None):  # pylint: disable=argum
-            from aiida.orm import Group
-
-            if current_autogroup is not None:
-                if not isinstance(current_autogroup, Autogroup):
-                    raise exceptions.ValidationError('`current_autogroup` is not of type `Autogroup`')
-
```
No, we manage internally the class, so this validation is just going to slow down storing (this is the performance improvement I was mentioning).
Anyway if the class is wrong because the user messed up, then the code would raise elsewhere. I'll keep it removed.

giovannipizzi commented on aiida/orm/autogroup.py:
```diff
@@ -32,148 +34,240 @@ class Autogroup:
+        queryb = QueryBuilder().append(
+            Group,
+            filters={
+                'or': [{
```
The logic is a bit more complex. Say the prefix is `a`. We anyway need to treat specially the case in which no group exists (neither `a` nor `a_%`), the case in which *only* `a` exists, and all other cases. The more complex query is to avoid that we also match labels like `abc`.

giovannipizzi committed:
Addressed comments by Sebastiaan.

In particular, the most important changes include:
- now the auto-group flag is called `--auto-group`, is
  a flag and is False by default
- only kept `--include` and `--exclude` options, checking
  typestrings and allowing to end with `%`. One benefit
  is that while reimplementing I replaced some `isinstance`
  with string comparisons, with potential further benefits.
- `--include` and `--exclude` are now mutually exclusive
- improved documentation of the main point of the issue

giovannipizzi commented:
Addressed all comments in a new commit. Let's see how many rounds of testing I'll need to do this time...

sphuber commented:
You should be able to run the docs and pre-commit also locally. For docs `make -C docs html` and for pre-commit `pre-commit run --all`. Then you won't have to push multiple times and depend on the CI

giovannipizzi commented:
Yes, thanks.
I know these two - for the docs, I was too optimistic and didn't think to have broken it.

For the pre-commit, the problem is that if I check only the problematic files, I get no error... (with `pre-commit run --files tests/cmdline/commands/test_run.py`).

I could run `--all-files` but it takes (on my PC) ~2 min spinning all CPUs at 100%.
Most problematic is the fact that if I run on all files I get errors that I don't see if I just check a single file, and even more weirdly I get errors in new parts of the code randomly popping out (e.g., I just run and I got this:
```
aiida/tools/ipython/ipython_magics.py
  Line: 37
    pylint: no-name-in-module / No name 'version_info' in module 'IPython'
  Line: 38
    pylint: import-error / Unable to import 'IPython.core'
    pylint: no-name-in-module / No name 'core' in module 'IPython'
```

I'll fix this (and I already fixed something similar earlier) but it's just very confusing.

BTW, a third test ([sqlalchemy, py3.5](https://github.com/aiidateam/aiida-core/pull/3650/checks?check_run_id=557868507#step:12:13)) failed because it couldn't upload to codecov... Did anybody already see it? Mentioning @csadorf for info

sphuber commented:
> BTW, a third test ([sqlalchemy, py3.5](https://github.com/aiidateam/aiida-core/pull/3650/checks?check_run_id=557868507#step:12:13)) failed because it couldn't upload to codecov... Did anybody already see it? Mentioning @csadorf for info

This I have also already seen multiple times. For me was also py3.5 but probably just coincidence. Seems like a transient connection problem. Would be nice if this could be couched in a retry or if it could not fail the test. But then again, the coverage probably depends on it.

sphuber commented on aiida/orm/autogroup.py:
```diff
@@ -7,173 +7,286 @@
+                    "'{}' has an invalid prefix, must be among: {}".format(string, sorted(valid_prefixes))
+                )
+
+            # If a % is present, it can only be the last character
```
Is this just a rule that we decide, or an intrinsic one?
The following entry points are perfectly valid:
```
aiida.data:plugin.base.int
aiida.data:plugin.complex.int
```
so why wouldn't `aiida.data:plugin.%.int` not be a valid filter?

giovannipizzi replied:
I was thinking for simplicity to only consider the most common case. However I have now lifted this requirement.

sphuber commented on aiida/orm/autogroup.py:
```diff
@@ -7,173 +7,286 @@
-        except AttributeError:
-            return []
+
+        Returns ``None`` if no exclusion list has been set."""
```
```suggestion
        Returns ``None`` if no inclusion list has been set."""
```

giovannipizzi replied:
Fixed.

sphuber commented on aiida/orm/autogroup.py:
```diff
@@ -7,173 +7,286 @@
+            # We should never be here, anyway - this should be catched by the `set_include/exclude` methods
+            raise ValueError("You cannot specify both an 'include' and an 'exclude' list")
+
+        the_class = node.__class__
+        if issubclass(the_class, ProcessNode):
+            try:
+                the_class = node.process_class
+            except ValueError:
+                # It does not have a process class - we just check the node class then, it could be e.g.
+                # a bare CalculationNode.
+                pass
+        class_entry_point_string = get_entry_point_string_from_class(the_class.__module__, the_class.__name__)
```
All of this can just be replaced with:
```
entry_point_string = node.process_type
# If there is no `process_type` we are dealing with a `Data` node so we get the entry point from the class
if not entry_point_string:
    entry_point_string = get_entry_point_string_from_class(node.__class__.__module__, node.__class__.__name__)
```

giovannipizzi replied:
Thanks

sphuber commented on aiida/orm/autogroup.py:
```diff
@@ -7,173 +7,286 @@
+            # Include all classes by default if nothing is explicitly specified.
+            return True
+        if include is not None and exclude is not None:
+            # We should never be here, anyway - this should be catched by the `set_include/exclude` methods
+            raise ValueError("You cannot specify both an 'include' and an 'exclude' list")
```
if this is true, then it is better to make this an `assert`

giovannipizzi replied:
Fixed.

sphuber commented on aiida/orm/autogroup.py:
```diff
@@ -7,173 +7,286 @@
+        # I import here to avoid circular imports
+        from aiida.orm.nodes.process import ProcessNode
+
+        # strings, including possibly 'all'
```
this comment is no longer true, correct?

giovannipizzi replied:
Correct. Removed.

sphuber commented on aiida/orm/autogroup.py:
```diff
@@ -7,173 +7,286 @@
+        If no group label prefix was set, it will set a default one by itself."""
+        return self._group_label_prefix
 
     def get_group_name(self):
```
Since we are anyway breaking backwards compatibility, should we rename this to `get_group_name`?

giovannipizzi replied:
I kept it for backward compatibility (the function is deprecated). As you noticed, I ended up removing other stuff that is not really deprecated, but this was internal logic anyway (and not working until now) so I don't think it's a problem.

sphuber reviewed (changes requested):
Thanks Gio. Unfortunately I still have some questions. Happy to discuss in person

giovannipizzi committed:
A few further comments addressed.

Most notable: now the % sign can be anywhere in the string
of included or excluded classes, and not only at the end.

giovannipizzi commented on aiida/cmdline/commands/cmd_run.py:
```diff
@@ -80,22 +108,26 @@ def run(scriptname, varargs, group, group_name, exclude, excludesubclasses, incl
+        warnings.warn('--group-name is deprecated, use `--auto-group-label-prefix` instead', AiidaDeprecationWarning)  # pylint: disable=no-member
+        if auto_group_label_prefix:
+            raise click.BadParameter(
+                'You cannot specify both --group-name and --auto-group-label-prefix; use --group-label only'
```
thanks, fixed

giovannipizzi commented:
All comments addressed and tests pass.

sphuber committed:
Merge branch 'develop' into fix_997_parallel_autogroups

giovannipizzi committed:
Minor pre-commit fixes after merge.

giovannipizzi commented:
This PR is cursed ;-) Fixed the problems after the merge, now ready to be merged!

sphuber commented:
Let the curse be gone! Thanks @giovannipizzi !
